packages/apps/plugins/plugin-thread/src/components/MessageContainer.tsx

```diff
@@ -0,0 +1,133 @@
+//
+// Copyright 2024 DXOS.org
+//
+
+import { DotsSixVertical, X } from '@phosphor-icons/react';
+import React, { forwardRef, useState } from 'react';
+
+import { type Message as MessageType } from '@braneframe/types';
+import { Surface } from '@dxos/app-framework';
+import { type SpaceMember } from '@dxos/client/echo';
+import { type Expando, getTextContent, TextObject } from '@dxos/echo-schema';
+import { PublicKey } from '@dxos/react-client';
+import { Button } from '@dxos/react-ui';
+import { TextEditor, useTextModel } from '@dxos/react-ui-editor';
+import { Mosaic, type MosaicTileComponent } from '@dxos/react-ui-mosaic';
+import { hoverableControlItem, hoverableControls, hoverableFocusedWithinControls, mx } from '@dxos/react-ui-theme';
+import { Message, type MessageBlockProps, type MessageProps } from '@dxos/react-ui-thread';
+
+import { useMessageMetadata } from '../hooks';
+import { THREAD_ITEM } from '../meta';
+
+type Block = MessageType.Block;
+
+const safeParseJson = (data: string) => {
+  try {
+    return JSON.parse(data);
+  } catch (err) {
+    return data;
+  }
+};
+
+const messageControlClassNames = ['p-1 min-bs-0 mie-1 transition-opacity', hoverableControlItem];
+
+const ObjectBlockTile: MosaicTileComponent<Expando> = forwardRef(
+  ({ draggableStyle, draggableProps, item, onRemove, active }, forwardedRef) => {
+    let title = item.name ?? item.title ?? item.__typename ?? 'Object';
+    if (typeof title !== 'string') {
+      title = getTextContent(title);
+    }
+
+    return (
+      <div
+        role='group'
+        className={mx(
+          'grid',
+          active === 'overlay' ? 'grid-cols-[min-content_1fr_min-content]' : 'col-span-3 grid-cols-subgrid',
+        )}
+        style={draggableStyle}
+        ref={forwardedRef}
+      >
+        <Button variant='ghost' classNames={messageControlClassNames} {...draggableProps}>
+          <DotsSixVertical />
+        </Button>
+        <div role='none' className={onRemove ? '' : 'col-span-2'}>
+          <Surface role='message-block' data={item} fallback={title} />
+        </div>
+        {onRemove && (
+          <Button variant='ghost' classNames={messageControlClassNames} onClick={onRemove}>
+            <X />
+          </Button>
+        )}
+      </div>
+    );
+  },
+);
+
+const TextboxBlock = ({
+  defaultContent,
+  onBlockDelete,
+}: { defaultContent?: string } & Pick<MessageBlockProps<Block>, 'onBlockDelete'>) => {
+  const [textObject] = useState(new TextObject(defaultContent));
```

Comment:
I'm a little confused by this. I think the prop should be a text object itself? These will come straight from the db as text objects in comments that we want to be live editable and going to string and back to text object seems likely to break something.